### PR TITLE
Tmi2 365/create mandatory question table

### PR DESCRIPTION
**Pull request by GavCookCO**
Adding CRUD functionality for Mandatory Questions. 

Have left some TODOs in the code that I will refactor out as we work through the other stories but for the sake of unblocking the sprint, the code that is here works fine.

### CODE REVIEW

**Review by a-lor-cab — DISMISSED** [review]

**Review by a-lor-cab — APPROVED** [review]

File: src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantMandatoryQuestionsController.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+        modelMapper.map(mandatoryQuestionDto, grantMandatoryQuestions);
+        if (mandatoryQuestionDto.getOrgType() != null) {
+            grantMandatoryQuestions.setOrgType(GrantMandatoryQuestionOrgType.valueOfName(mandatoryQuestionDto.getOrgType()));
+        }
+        if (mandatoryQuestionDto.getFundingAmount() != null) {
+            grantMandatoryQuestions.setFundingAmount(new BigDecimal(mandatoryQuestionDto.getFundingAmount()));
+        }
+        grantMandatoryQuestions.setId(mandatoryQuestionId);
+        grantMandatoryQuestions.setStatus(GrantMandatoryQuestionStatus.IN_PROGRESS);
```
Comment (iaincooper-tco) [inline src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantMandatoryQuestionsController.java:95]: I don't think we want to set it to In Progress any time we update.
The user can complete their mandatory questions, and change the answers later and that shouldn't set the status back to in progress.
There will be more complicated logic added here later so its fine to put a TODO for now.

File: src/main/java/gov/cabinetoffice/gap/applybackend/service/GrantMandatoryQuestionService.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        return grantMandatoryQuestion.get();
+    }
+
+    public GrantMandatoryQuestions createMandatoryQuestion(GrantScheme scheme, GrantApplicant applicant){
+
+        // TODO we probably don't need to ask if it exists and instead react if the result is null
+        if (doesMandatoryQuestionAlreadyExist(scheme, applicant)) {
+            log.debug("Mandatory question for scheme {}, and applicant {} already exist", scheme.getId(), applicant.getId());
+            return grantMandatoryQuestionRepository.findByGrantSchemeAndCreatedBy(scheme, applicant).get(0);
+        }
```
Comment (iaincooper-tco) [inline src/main/java/gov/cabinetoffice/gap/applybackend/service/GrantMandatoryQuestionService.java:40]: definitely a valid use case for both here
the create method should error out if it's trying to create a mandatory question which already exists
however in our usage, there is a benefit to us doing a create-or-retrieve as that is the process we'd want in the frontend.
might be worth having two methods, a create method which errors out if it already exists, and a create-or-retrieve method which uses the previous one and does the retrieval if it errors.

**Review by iaincooper-tco — CHANGES_REQUESTED** [review]

File: src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantMandatoryQuestionsController.java (hunk 1 above)
Comment (a-lor-cab) [reply src/main/java/gov/cabinetoffice/gap/applybackend/web/GrantMandatoryQuestionsController.java:95]: added an if to change the status only if it was NOT_STARTED

File: src/main/java/gov/cabinetoffice/gap/applybackend/service/GrantMandatoryQuestionService.java (hunk 2 above)
Comment (a-lor-cab) [reply src/main/java/gov/cabinetoffice/gap/applybackend/service/GrantMandatoryQuestionService.java]: created a todo

**Pull request merged**